rpa_queue.h

```diff
@@ -39,10 +40,76 @@
  */
 
 /**
- * opaque structure
+ * @brief Opaque structure representing a thread-safe circular queue.
  */
 typedef struct rpa_queue_t rpa_queue_t;
 
+/**
+ * @struct rpa_queue_t
+ * @brief Opaque structure representing a thread-safe circular queue.
+ *
+ * This structure defines a thread-safe circular queue that supports concurrent access
+ * by multiple threads. The queue uses a fixed-size array to store elements and is designed
+ * to handle both producer and consumer threads efficiently.
+ */
+struct rpa_queue_t {
+  void **data;                  /**< Array to store elements */
+  volatile uint32_t nelts;      /**< Number of elements in the queue */
+  uint32_t in;                  /**< Next empty location in the queue */
+  uint32_t out;                 /**< Next filled location in the queue */
+  uint32_t bounds;              /**< Maximum size of the queue */
+  uint32_t full_waiters;        /**< Number of threads waiting on a full queue */
+  uint32_t empty_waiters;       /**< Number of threads waiting on an empty queue */
+  pthread_mutex_t *one_big_mutex;/**< Mutex for controlling access to the queue */
+  pthread_cond_t *not_empty;    /**< Condition variable for signaling non-empty queue */
+  pthread_cond_t *not_full;     /**< Condition variable for signaling non-full queue */
+  int terminated;               /**< Flag indicating whether the queue is terminated */
+};
+
+/**
+ * @brief Macro to check if the rpa_queue_t is full.
+ *
+ * This macro checks if the number of elements in the queue is equal to the maximum
+ * size of the queue, indicating that the queue is full.
+ *
+ * @param queue Pointer to the rpa_queue_t instance.
+ * @return 1 if the queue is full, 0 otherwise.
+ */
+#define rpa_queue_full(queue) ((queue)->nelts == (queue)->bounds)
```

Comment:
Thanks for the contribution, but can you do this with functions rather than macros? That would allow keeping the structure private (hidden in the `.c` file).

Reply:
Done :)